### PR TITLE
Performance enhancements for InfluxDB with large graphite metric series

### PR DESCRIPTION
**Pull request by pkittenis**
This PR can now be merged. It is almost a re-write with only the `statsd` timers remaining from the parent project.

Requirement for the re-write was good performance with InfluxDB being used as a drop in replacement for Graphite's `carbon-cache`. In that use-case InfluxDB has one unique series per individual graphite metric path.

On large graphite installations, that results in tens of thousands or more worth of series in InfluxDB which the original `graphite-influxdb` struggles to cope with. With these changes, looking up graphite metric paths is a sub 200ms operation in most cases, compared to ~5-10 seconds to several minutes with the original `graphite-influxdb`. 

Memcache integration brings load on InfluxDB down further and ensures than re-indexing happens from series list from memcache.

Even without memcached integration being used, metric path lookups perform the the same though re-indexing is bottlenecked at getting series list from InfluxDB in that case.

Changes:
- Schema-less design based on query time interval. No manual schema configuration needed. Interval to use for a given query time range is configurable
- Python 3.4 and 3.5 support with tests
- Configurable aggregation functions for metric regex pattern. Equivalent of `storage-aggregations.conf` in `carbon-cache`
- Configurable retention policies to use based on query time interval
- Elasticsearch integration removed - in memory index used for path finding instead
- Memcached integration added as an extras option (not enabled by default)
- Memcache caching of series list on configurable TTL
- Memcache caching of metric data with TTL matching query time interval
- Series list pagination
- Removed regex matching in InfluxDB in favour of in memory index and Graphite glob matching from graphite-api package
- Background series list loading into memcache so that index building uses cached series list
- Node / metric path lookups use in memory index only
- Many more integration tests added for edge cases in finding nodes, new configuration options et al.
- Removed graphite-web support in favour of Graphite-API. Graphite-Web is very slow when used with InfluxDB as it cannot do multi-fetch. Support for it may be re-added with little effort but there are no tests for graphite-web and therefore no guarantees that it will work or not break in future
- Package refactoring. Classes in `graphite_influxdb.classes`, utils in `graphite_influxdb.utils`, constants in `graphite_influxdb.constants`

Not implemented:
- InfluxDB tags are not supported. Suggestions on ways to support tags without breaking Graphite API compatibility while (the hard bit) retaining good performance are most welcome.

Testing, comments, contributions are all welcome!

TODO list prior to merging:
- [x] Tests for parent nodes tree traversal
- [x] Configurable series loader interval
- [x] Removal of unused in-process node caching
- [x] Update readme
- [x] Index load on startup and save on build using existing Graphite-API search index configuration

Thanks also to @mbell697 and @r3cgm for their contributions and testing.

Example of wild card query with over 2M series stored in InfluxDB with these changes:

multi_wildcard_query

Compared to original `graphite-influxdb` without and with memcache respectively:

fc2ea104-9e63-11e5-9a23-2f49512c478d

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,115 +1,223 @@
...
+.. image:: https://coveralls.io/repos/pkittenis/graphite-influxdb/badge.png?branch=master
+  :target: https://coveralls.io/r/pkittenis/graphite-influxdb?branch=master
 
-Install and configure using docker
-----------------------------------
 
-Using docker is an easy way to get graphite-api + graphite-influx up and running.
-See https://github.com/Dieterbe/graphite-api-influxdb-docker which provides
-a container that has all packages installed to make maximum use of these tools.
+This project is a fork of the excellent `graphite_influxdb <https://github.com/vimeo/graphite-influxdb>`_ finder. Many thanks to Vimeo and the author for open sourcing that work.
```
Comment (Dieterbe) [inline README.rst]: Obviously changes like this have to be undone to merge your work in.

File: README.rst [hunk 2]
```diff
@@ -1,115 +1,223 @@
...
+------------------
+
+In this project, no per series schema configuration is required, as with `InfluxDB`_.
+
+Calculated intervals
+--------------------
+
+An interval, or step, used to group data with is automatically calculated depending on the time range of the query.
+
+This mirrors what `Grafana`_ does when talking directly to InfluxDB.
```
Comment (Dieterbe) [inline README.rst]: Graphite and grafana use maxdatapoints variable for this. We should too, probably?

**Comment by Dieterbe** [comment]
Glad to see you got memcache to perform well.when I tried to use it as a cache it was too slow. This looks pretty good, but why remove ES support? Can't we support both ES and memcache?

**Comment by mbell697** [comment]
Thanks for all your work on the @pkittenis, I'll test this out in the next few days. 

I am still a bit concerned about the lookup time. Just one of our dashboards can trigger dozens of metric lookup queries to influx and I found that to be the biggest bottleneck, if they each take anything like ~250ms I imagine we'll still be using Varnish as a cache for them. 

It seems to me that caching every series directly is still doing a lot of repeated work for every query and using ES to do something simple like this is massive overkill. Have you considered storing a serialized form of the tree data structure I recommended here: https://github.com/pkittenis/graphite-influxdb/issues/13 ? memcached can be used to share the serialized index data structure between processes and to handle atomic updates of it. You'd only need to pull the data structure from memcached periodically, roughly as often as it is updated. Assuming the sharing and updating of the data structure can be solved the query performance should be extremely good.

File: README.rst (hunk 2 above)
Comment (pkittenis) [inline README.rst]: Maxdatapoints is a parameter to Graphite API and is honoured by Graphite API - not applicable to Graphite API storage plugins.

**Comment by pkittenis** [comment]
Thanks for testing @mbell697. Yes, a cache in front of influxdb API would protect it from multiple identical requests and furthermore a webapp level cache can be enabled on the Graphite-API front by various means. This module is a plugin to an existing webapp - it therefore has caching specific to its interactions, namely interactions with InfluxDB.

Building a tree and storing it in memcached is what background series loader and the new branch algorithm are doing. Comments on why not store the tree as a (single) object are at  pkittenis#13 , namely such a tree would be too big to fit in a single memcached value and would cause locking issues when updating it. Same issue with storing entire series list in memcache/in python process/anywhere else without pagination.

The tree would have to be updated by only one operation at a time which would make locking/unlocking tree updates a bottleneck for every `metrics/find` operation and every operation that needs to do a tree update.

OTOH, storing and caching levels of the tree separately (ie, each branch), means that each branch can be updated independently of each other with no bottlenecks.

Storing an indexed tree `somewhere` is the same solution as the existing elasticsearch support and could be used instead of this if that solution is preferred. 

In our tests, the memcached approach is far lighter on resources and better performing. YMMV.

File: README.rst (hunk 1 above)
Comment (pkittenis) [inline README.rst]: Yup, 'update readme' todo task above :)

**Comment by pkittenis** [comment]
| why remove ES support? Can't we support both ES and memcache?

ES is massively expensive to run and could not be feasibly run on the same hardware as InfluxDB as they both require a lot of RAM to work well.

Memcached is a light weight in memory key/value store and uses roughly 250MB of RAM or  to store data for an InfluxDB instance on the same machine taking up ~32GB of memory.

If it can be shown that ES can match both performance and hardware requirements of the memcached solution then great. Otherwise, what is the benefit of having such integration.

Both could be supported, yes. Given that there are no instructions on how to generate ES data for influxdb series list, no tests for it, no automated updating of ES index from influxdb data, and that memcached seems to out perform it with lower hardware requirements, that support is not something I will be working on.

**Comment by mbell697** [comment]
I think we're missing each other so let me explain in full:

* Periodically load the full set of series names from influx in a background thread. To allow only a single process to be the 'updater' at a time a simple lock can be taken out in memcached. This update can happen periodically, say every 15 minutes. 

* The background thread constructs a k-ary tree of series names and serializes it to a single memcached key. As for the concern over the size of this, I just pulled 265k series from our production data and serialized the index as json in a nested array structure. The index is 691KB gzipped, which is how memcached will store it and it can be passed to memcached already gzipped to save on network IO if that is a concern. Uncompressed it's 6.1MB. This is far from the most space efficient serialization and I would expect non-linear space usage as the # of metrics increases but that will vary with the metric naming structure. I don't expect index size to be a significant problem and don't think memcached will impose any major limitation on it. 

* Each process has a background thread that periodically loads this index and replaces the current in memory index with it. The de-serialization can happen in the background thread so all that is needed is a simple and short lived write lock when the new index replaces the old index in memory.

* All metric queries hit this index with a simple search algorithm. There is no need to hit memcached when handling a request. There will also never be a 'cache miss' in this organization other than perhaps the very first time the system comes online and builds the very first index. You will of course have stale data for ~15m or however long it is between index refreshed but that is an issue with any approach. With the various example patterns I've tested with this I get 2.8k-38k queries per second against the index so performance will be dominated by the rest of the system, i.e. graphite-api request processing. 

I updated the mockup at https://gist.github.com/mbell697/f1d4c9a01a8617e5b50b with following:

* Improved insertion performance by storing the children in a map by name. It now takes ~1.9s to build an index with 265k series names being read from a file.

* Added serialization / deserialization from JSON. It takes < 2s for each operation with an index of 265k names and writing/reading to a file on disk. 

* Couple minor performance improvements on the lookup although it still does some very dumb things. Query rate is anywhere from 2.8k - 38k queries per second depending on depth / complexity of the query. e.g. the last query in the benchmark returns 44 series.

As mentioned before I wrote this in Ruby just for the sake of a quick mockup but can translate to Python. I would expect the performance to be about the same.

**Comment by pkittenis** [comment]
| Periodically load the full set of series names from influx in a background thread

This is what series loader currently does.

| The background thread constructs a k-ary tree

Background thread instead stores each branch of the tree in separate key values.

| Each process has a background thread that periodically loads this index

This could be added to avoid having to hit memcache, yes.

| All metric queries hit this index with a simple search algorithm

This already happens on the memcache data.

The only difference from current implementation and the proposal seems to be how the index is stored and whether or not it is held in python process memory or queried from memcache each time.

Would say the former is an implementation detail and the latter a potential improvement which would at least save having to hit memcache. Personally I find having a [small node finding algo](https://github.com/pkittenis/graphite-influxdb/blob/master/graphite_influxdb/classes/finder.py#L142) to be much simpler than implementing yet another indexed tree. Others may disagree :)

The indexed tree approach also assumes all series can be loaded from influxdb at the same time, which is simply not true. InfluxDB unfortunately imposes a maximum limit of 10,000 rows returned from one request, including series lists. Is is therefore not possible to retrieve an entire series list from influxdb once it passes >10k.

Given that there used to be in python process caching and how poor python is with memory management of large lists, don't think re-adding that is a good option. Would love to see some tests on this though, ie how much performance is gained by not having to hit memcache. The [relevant code is here](https://github.com/pkittenis/graphite-influxdb/blob/master/graphite_influxdb/classes/finder.py#L128) which can be overridden by an in memory cache of each key.

With our live data a full series list of json serialised data for ~30k series from Influxdb takes 1.1MB compressed. Default max key value for memcache is 1MB. 

Obviously this does not match your data and may be much more than most people, still, exceeding memcache key length is not just a possibility. Once the tree becomes large enough it simply won't fit.

Not disagreeing that that is a good approach, but given the data that the live tree that we have will not fit in memcache's default settings, previous tests showing python using a lot of memory to store the same tree in process and the impossibility with current influxdb to not have cache misses at all, I'm afraid would have to see real data showing a significant benefit to consider implementing that over the current implementation.

**Comment by mbell697** [comment]
> With our live data a full series list of json serialised data for ~30k series from Influxdb takes 1.1MB compressed. Default max key value for memcache is 1MB.

We're still missing each other, I'm not at all proposing storing all raw series names in memcached, I'm proposing storing an index of them:

`curl -G 'http://prod-influxdb-b01:8086/query' --data-urlencode "db=graphite" --data-urlencode "q=SHOW SERIES" > influx_full_series.json`

Produces a 65MB slug of JSON for our data which takes about 30 seconds for influx to generate and send through a VPN to my machine. But the index I'm proposing for that data is only 691KB which is all that needs to be stored in memcahed. There is no row limit on `show series` at the moment, it returns all series, unless that has changed very recently. There is a row limit of 10k on the `show measurements` query and I would still paginate the loading to be nice on influx if possible, but that is an implementation detail.

> Obviously this does not match your data and may be much more than most people, still, exceeding memcache key length is not just a possibility. Once the tree becomes large enough it simply won't fit.

Even if you assume linear index growth with # of metrics, which isn't realistic, you'd need something like 50 million metrics to exceed the maximum value size of memcached which is 128M last I looked. At that size there are plenty of other issues that would come up before this one so I would largely consider it a non-issue.

**Comment by mbell697** [comment]
I pulled the 0.5.2b6 release to test today, I assume that matches up with this PR?

Unfortunately I wasn't able to get very far with it as it was killing influx. 

Queries like: `SHOW MEASUREMENTS WITH MEASUREMENT =~ /^statsite\.some\.metric\.name\.received/ LIMIT 2000 OFFSET 0` appear to be insanely slow, that is a leaf metric so only returns 1 single series and takes 36 sec to execute. A single dashboard triggers several dozen of these queries so I was never able to get much of anything to load due to request timeouts. 

For comparison: `SHOW SERIES FROM \"graphite\".\"default\"./^statsite\.some\.metric\.name\.received/` executes in about 1 second. Actually `SHOW SERIES` dumping all series data (65MB of json) executes faster than the `SHOW MEASUREMENT` query returning a single series above. 

Clearly this is a problem somewhere in Influx but it's blocking any further testing on my end. 

I did notice the background series loader seemed to stop around offset 10000 but it's unclear if that was related to slowdowns caused by the above queries or caused by something else.

**Comment by pkittenis** [comment]
From release notes:

| Series list now uses pagination by default and caching is per page. Note that limit queries (for pagination) *do not apply in InfluxDB versions prior to 0.10b1*.

**Comment by pkittenis** [comment]
| I'm not at all proposing storing all raw series names in memcached, I'm proposing storing an index of them

What is stored in memcache is a compressed python list of series names generated from influxdb data, not influxdb data itself. 

When the proposal is demonstrated with significant benefits over the current implementation it can be taken up.

**Comment by pkittenis** [comment]
Ok, so have found a few new things.

 * show series does not have 10k limit
 * show series does not allow regex - it accepts regex but gives full list regardless so it cannot be used for regex based queries
 * show series now also honours limits as of 0.10

With our live data (~350k series), memcache can store roughly 150k compressed series in its default 1MB max value.

Have made [incomplete changes](https://github.com/pkittenis/graphite-influxdb/tree/show_series) that use show series instead of show measurements to get a full series list and that solves the cache miss problem. Feel free to test with that branch @mbell697 - use latest 0.10-rc1 influxdb.

However, finding branches can now take a while on such a large series list - ~3s for ~350k series. So it looks like an indexed tree is the way to go. When that will be, don't know. Resources for implementing it would be appreciated.

**Comment by mbell697** [comment]
> So it looks like an indexed tree is the way to go, however, it will have to be stored per branch to avoid hitting memcache limit as above.

I doubt you'll hit the 1MB limit with 350k, the index does not store every full series name but the final size will depend on your metric structure. You can also always just up the value size limit in memcached or use Redis / File storage. 

I ported the index implementation to python last night, it performs roughly the same as the Ruby implementation. It's likely not very pythonic and isn't tested much but it should serve as an example:https://gist.github.com/mbell697/00bbc559e504063c49b2

**Comment by pkittenis** [comment]
The index contains links to objects containing list of nodes, all of which needs to be serialised and stored in memcache.. In both cases all series names need to be stored somewhere, the where and how is again an implementation detail. Unless it's possible to somehow magically produce a list of series from an index alone without also storing what the index is pointing to. 

Yes, it depends on the hierarchy - greater number of unique paths will produce bigger indexes. In our case our live data has several non-identical hierarchies and multiple root branches which produces a very big unique list of series.

The example code also does not seem to cater for multiple wildcards in the same path and still uses [python regex to do the searching](https://gist.github.com/mbell697/00bbc559e504063c49b2#file-index_test-py-L80). What would happen if `query(Query('*')` were run? How about query(Query('statssite.*.*.*.cpuusage')`. Won't a regex match be run on every child under root branch? Will that not run regex matches on thousands of series? This is the reason the original implementation was so slow btw.

In any case, you could give it a go and show some results since you have example code already :)

**Comment by mbell697** [comment]
> The index contains links to objects containing list of nodes, all of which needs to be serialised and stored in memcache

No, only the index needs to be stored :)

> Unless it's possible to somehow magically produce a list of series from an index alone without also storing what the index is pointing to.

That is exactly what the example code does. There is no magic about it, metric names are constructed by traversing the index tree.

> What would happen if query(Query('*') were run?

It returns the branches under the root. 

```
> store.query('*')
[ { 'is_leaf': False, 'metric': 'apps'},
  { 'is_leaf': False, 'metric': 'servers'},
  { 'is_leaf': False, 'metric': 'services'},
  { 'is_leaf': False, 'metric': 'statsite'}]
```

> How about query(Query('statssite.*.*.*.cpuusage') Won't a regex match be run on every child under root branch? Will that not run regex matches on thousands of series? This is the reason the original implementation was so slow btw.

At present it would match the regex against each node in those three levels but optimizing away `*` queries to just return all children is a one line change (added to the gist) and the other style of glob expressions that graphite supports will require matching no matter what the structure is. As for now many regexs will be processed, it depends on the metric structure, keep in mind that matching is done at each level of the tree so your never matching against all leaf paths under a high level parent. 

It's probably possible to remove other regexes usage if that is actually a bottleneck, not sure how fast Python's regex engine is but Ruby's is often about as fast as other approaches even for simple regexs, such as the exact match done in the example code. In most common cases I would expect the number of regex tests required to be on the order of the # of series returned so it shouldn't be problematic. 

If your curious about specific situations, it's probably easiest to test this against your production data. `index_test.py` expects an input file named `names_only` which is a file with one metric name per line. This can be generated by grabbing a full `show series` response and transforming it with `jq` or however you wish.

**Comment by pkittenis** [comment]
| That is exactly what the example code does. There is no magic about it, metric names are constructed by traversing the index tree.

So index does contain series names.. :) Anyway, I agree it's a good idea and should be done. It won't be included as part of this PR.

**Comment by pkittenis** [comment]
Hi @Dieterbe, do you want django/elasticsearch removals to be re-added for this PR?

For django, can re-add the imports but I will not have the time to update django configuration parsing for the new/changed config values nor test it works correctly beyond imports.

For elasticsearch, can add it back in as it was but will not be able to test it works. If you're happy with performance of this PR with memcache enabled ES can be left removed.

**Comment by Dieterbe** [comment]
re ES, there def. is/was a use for that, for people who also use carbon-tagger or graph-explorer, as they will populate the index (https://github.com/vimeo/graphite-influxdb#using-elasticsearch-as-an-index) it is possible that only very few people use it
re django, I have no idea what that is about.
but the larger problem is, i'm not using this project anymore right now.  I may start using it again later when i give influx 0.10+ another chance, but i'm not sure when that would be. so i also have no motivation to review or test your PR, or even maintain this project, at this point in time.

so i'm not sure what's the right way forward. some ideas:
* make you maintainer of this project and let you do whatever you want.
* mark this project as unmaintained until further notice, update readme to point to your project?

**Comment by pkittenis** [comment]
Hi,

Thanks for the info on ES - can add that back in.

Regarding ongoing maintenance, that was actually partly why this PR was raised - to gauge if the project is still being maintained :)

Making me maintainer of this github project would require access to the corporate account so that does not sound feasible. I'm not sure vimeo would like someone that doesn't work for them to have access to the corp account :)

Be happy to take up maintenance of the `graphite-influxdb` pypi project on the other hand. The fork is only currently published on github, is not on the pypi index at all. My username on pypi is the same as my github account.

If you would like to mark this project as unmaintained and link to the fork that's fine with me. Completely up to you of course.

**Comment by mbell697** [comment]
To catch back up with this, I did a full rewrite using the index structure I discussed earlier in this thread, that code is available here: https://github.com/swoop-inc/graflux, it likely doesn't solve all the possible issues but solves ours.

**Comment by pkittenis** [comment]
Thanks, a PR against https://github.com/pkittenis/graphite-influxdb/ or this parent project would be useful.

Furthermore, while I'm no lawyer and 'full rewrite' without references to any source material claims would only need proving in a court of law, at least some code and functionality in there does appear to be at least partly based on the parent project and the fork, both of which are licensed under the Apache license, which if code is derived from would not be allowed to be re-licensed under the MIT license without written consent from copyright owners..

To name just a few examples, [this code here](https://github.com/swoop-inc/graflux/blob/master/graflux/query_engine.py#L40-L41) looks [almost identical to this code from the fork](https://github.com/pkittenis/graphite-influxdb/blob/master/graphite_influxdb/utils.py#L146-L147) even using same internal variable names, so [does this code](https://github.com/swoop-inc/graflux/blob/master/graflux/query_engine.py#L37) look identical [to this](https://github.com/vimeo/graphite-influxdb/blob/master/graphite_influxdb.py#L301) from the parent project, ditto for [steps/deltas and aggregation function configuration](https://github.com/swoop-inc/graflux/blob/master/config/graphite-api.yaml#L20-L45) [compared to the fork](https://github.com/swoop-inc/graflux/blob/master/config/graphite-api.yaml#L20-L45),  down to the commented out under 1 hour intervals..

**Comment by mbell697** [comment]
I'm more than happen to provide any attribution you would like and if you have a strong feeling that there is enough similarity I can relicense under APL. As the scope of a project like this is very small and it converts between two fixed APIs it is likely for the implementations to be similar at points, particularly at API boundaries.

It's unclear to me that a PR makes sense as the projects structures are very different, i.e. a PR would replace every code file in the project and there isn't 1:1 feature parity. Certainly one of the reasons for using an MIT license was to make it easier for you (or anyone else) to use all or parts of the code without concern for license compatibility.

**Comment by pkittenis** [comment]
Hello again,

PR is now good to go as far as I know. Path lookups are now index based and perform the same regardless of any caching, though memcache integration is still highly recommended. 

This version has been running very well for us with ~1M series in the database.

Do keep in mind memory use can be very high with that number of series in InfluxDB, however.

The PR is lacking index saving/loading so index is always re-built at startup. To re-iterate, memcached integration is highly recommended to be enabled, also in the readme.

It supports influxdb versions >=0.9.2, including 1.0.

**Comment by pkittenis** [comment]
Thank you for the heads up - this version [handles that](https://github.com/vimeo/graphite-influxdb/pull/49/files#diff-ca9f443e9be9c2770ed8365ce5fa6b3eR350) by not specifying a default RP at all and letting influxdb use the default configured on the database, unless retention policies are setup in plugin configuration, per readme file.

That way if the default retention policy is changed either by user or in influxdb code, the plugin will still work without any configuration or code changes.

**Comment by pkittenis** [comment]
Not sure if this project is still maintained - there has been no response to this PR in months. 

Have published the code in this PR plus some more features to Pypi and Github as [InfluxGraph](https://github.com/InfluxGraph/influxgraph) to allow people to use it easily without having to manually install some PR or a forked repo manually.

Though I do not like to advertise on other project's pages in this case the project is the same under a different name so hoping that is ok. Happy to remove this comment if there is/will be continued maintenance of this project :)

**Comment by Dieterbe** [comment]
This project is effectively unmaintained.
FWIW I gave up on influxdb and have been working on an alternative storage backend for graphite which has given us better results. (see https://github.com/raintank/metrictank)

I pushed a deprecation notice: https://github.com/vimeo/graphite-influxdb/commit/56e4aeec57f39c90f14f3fbec57641b90d06c08b

Have fun maintaining yours ;) Let me know if i can do anything else.

**Comment by pkittenis** [comment]
Thank you for the update and notice, closing this though do feel free to merge in the changes if you wish :)

That project you mentioned looks interesting, are there any demos showing its use from a client's perspective?

For your benefit and possibly others, it looks like our goals are aligned in having a drop in replacement for the (frankly, horrendous) carbon-cache while maintaining backwards compatibility with the Graphite API. I would agree Influx has moved to an API which is not easily compatible with Graphite's and particularly for discovering metrics and supporting wild card queries is a bit of a step (or two) backwards, certainly from a client's (Grafana and people using it) perspective. In my view its performance, features, in particular support for high resolution data, and low resource requirements more than make up for these limitations, speaking as someone with extensive experience in running and developing code for both Cassandra and Influx and with no affiliation with either.

Graphite-influxdb/InfluxGraph does provide a drop in replacement while ensuring Graphite is a first class citizen, both for existing clients using Graphite protocol and new clients using native Influx protocol, and gets particularly good performance with templating enabled in the storage finder and either graphite input with the same templates or native data in Influxdb. Maturity wise it has proven very stable and has been in heavy use for some time now, it should be considered a release candidate.

There are a few example configurations on the [wiki](https://github.com/InfluxGraph/influxgraph/wiki/Telegraf-default-configuration-template) if you'd like to try it out, at least in our write load skewed work load we get very impressive performance with many millions of metrics in the DB.

That said, I am by no means wedded to it, it came about from real world requirements and lack of alternatives as I imagine is the case with you, and if Cassandra and the above webapp perform better perhaps resources are best allocated there.

Thanks for reading :) You may reach me at the address in the project's setup.py if you'd like to discuss further.

**Comment by Dieterbe** [comment]
> That project you mentioned looks interesting, are there any demos showing its use from a client's perspective?

it basically just behaves like graphite. you can use the carbon input (also supported kafka and metrics2.0), we have a finder plugin for graphite-api (and soon for graphite-web) so you can query it like graphite. You can also query it directly (see https://github.com/raintank/metrictank/blob/master/docs/http-api.md) though it doesn't have the graphite processing functions built-in yet.  We aim to incoporate those so that at some point we don't require any python code anymore.
If you want to try it out, see https://github.com/raintank/metrictank/blob/master/docs/quick-start-docker.md

> and if Cassandra and the above webapp perform better perhaps resources are best allocated there.

FWIW metrictank and influxdb are both based on the same go-tsz library. performance and disk efficiency should be comparable.

**Comment by pkittenis** [comment]
> performance and disk efficiency should be comparable.

Was referring to cassandra performance, not the Go based API. In our tests, though they were a while ago now and with likely a different schema than what you are using, single node performance of cassandra as a metrics backend for Graphite was quite poor. Sure, it scales linearly as more nodes are added but if it requires say 100 nodes to match influx's single node performance..

Would be interested in some large scale ingestion test results if it is ready for that at this point.

**Comment by Dieterbe** [comment]
Yeah when using Cassandra the data model matters. Many solutions store points directly in cassandra, we store compressed chunks of typically 30min ~ 6hour in it as data values. We found this to be a huge (about 8x difference ) in performance/load on Cassandra. I have benchmarking on my to-do, will tackle that in the near future.
